### DISCUSSION TITLE
Increase the frequency of C++ API

### DISCUSSION THREAD

**Discussion by ClarkZha** [microsoft/AirSim#3859] [Support Q&A]
### What's your question?

How to increase the frequency of C++ API, specifically client.simSetVehiclePose().

### Include context on what you are trying to achieve

I am connecting an external drone dynamics simulator to AirSim+Unity. The external dynamics calculates the pose of the vehicle and send it to AirSim via the client.simSetVehiclePose(). Yet significant delay is observed as the client.simSetVehiclePose() is running at 50Hz. 

I cannot bypass the problem via setting the clock-time as the external physics runs with a hardware in the loop. 

Any hint/suggestion is much appreciated!

#### Context details

AirSim Version/#commit: 1.4.0
UE/Unity version: Unity 2019.3.12.f1
OS Version: Ubuntu 18.04

### Include details of what you already did to find answers
Related discussion about using external physics with Unity: #3385, #1828, #3203
Related discussion about API: #3049, #2118

**Comment by ClarkZha** [comment]
To add a little bit explanation to the discussion:

I made an experiment that moves the vehicle with `simSetVehiclePose()` and read the pose `simGetVehiclePose()` in a same loop that runs at 500Hz. Then I plot out the pose I set and I read as follows:

![delay](https://user-images.githubusercontent.com/39609430/123183904-68be4a80-d447-11eb-9e69-6c015b38c479.png)

I am seeing a 0.02 second delay between when the pose was set and when it is actually updated in the airsim. Moreover, I am seeing the pose being updated at a rate of 100Hz. I am wondering if there is anything I can do to decrease this delay time and increase the pose update frequency.

**Comment by skykim0609** [comment]
I am using AirSim for drone simulation, and I speculate that my issue is related to this. I am using airsim ros node and observed significant delay in multirotorrpcclient::getMultirotorState(). Perhaps the reason behind this is that somehow rpc::client::call() function(which also lies behind simSetVehiclePose() function) requires very long time(20ms in my case) to process, which is undesirable behavior.

**Comment by ClarkZha** [comment]
> I am using AirSim for drone simulation, and I speculate that my issue is related to this. I am using airsim ros node and observed significant delay in multirotorrpcclient::getMultirotorState(). Perhaps the reason behind this is that somehow rpc::client::call() function(which also lies behind simSetVehiclePose() function) requires very long time(20ms in my case) to process, which is undesirable behavior.

Yes I think what you experienced is related to what I observed. Can you please provide more background as this might be helpful with pinning down the problem?  : )

Are you using Unity/Unreal? At what rate are you calling multirotorrpcclient::getMultirotorState() in your ros node and what is the frequency your node is spinning at?

**Comment by skykim0609** [comment]
> > I am using AirSim for drone simulation, and I speculate that my issue is related to this. I am using airsim ros node and observed significant delay in multirotorrpcclient::getMultirotorState(). Perhaps the reason behind this is that somehow rpc::client::call() function(which also lies behind simSetVehiclePose() function) requires very long time(20ms in my case) to process, which is undesirable behavior.
> 
> Yes I think what you experienced is related to what I observed. Can you please provide more background as this might be helpful with pinning down the problem? : )
> 
> Are you using Unity/Unreal? At what rate are you calling multirotorrpcclient::getMultirotorState() in your ros node and what is the frequency your node is spinning at?

I am currently using AirSim 1.5.0 with Unreal 4.25 version, and I tried calling getMultirotorState function with 100Hz frequency, because I wanted to get IMU topic in at least 100Hz frequency.(This is done by modifying update_airsim_control_every_n_sec parameter in airsim_node.launch to 0.01) However I observed that the upper limit of topic frequency is about 47Hz, due to time required to process getMultirotorState().

**Reply by xxEoD2242** [reply]
Hey @skykim0609 

You are correct about this processing time. This process is running in a thread and this can be slowed down by the rest of your processes. What I would highly recommend is removing some of the calls after the get state call. In my experience, what's causing the slow down is how long it takes to actually build and publish the messages from the state, not the state call itself.

While the ROS wrapper is helpful, it is not written for high performance. We have worked to extensively rework our own version to meet our needs. Unfortunately, that wrapper is integrated very tightly with our Autonomy stack and wouldn't be useful for you to look at. In general, there is far too much processing going on in the ROS wrapper, so I would look to remove what you can i.e. turn off lidar or reduce the number of images you are getting.

**Comment by zimmy87** [comment]
We have checked in the code and the pose is being updated when the API is being called, so the update frequency is the msgpack RPC response frequency. You will need to figure out how to increase that frequency.

**Reply by ClarkZha** [reply]
Thank you zimmy for looking into this. To add a little bit more background to the discussion:

My test process: 
I set the Unity update rate to 500Hz. Meanwhile, in each main loop (also running at 500Hz): 
1. rpc client calls setpose API
2. rpc client calls getpose API 
3. plot out both results. 

I am observing: 
1. The getpose API return value changes at a rate at 100Hz, instead of 500Hz.
2. There is a consistent 20ms delay (indicating something running at 50Hz) between when I set the pose and when the get pose returns the same value. 

Does this matches what you are seeing/expect from the current RPC code?

I am familiar with the code I wrote and the C# code on the Unity side, but I am still learning RPC. It will be really helpful if you can provide some comments to help me understand the problem better. Specifically,
> update frequency is the msgpack RPC response frequency

Is this "msgpack RPC response frequency" what determines how often the setpose and getpose API goes through the RPC bridge, i.e., the frequency at which rpc::client::call() is triggered? 

Thank you very much!

**Reply by jonyMarino** [reply]
Hi @ClarkZha! Were you able to increase the rpc frequency?

**Reply by ClarkZha** [reply]
> Hi @ClarkZha! Were you able to increase the rpc frequency?

Thanks @jonyMarino ! Originally I thought the delay was caused by RPC as I find a 50Hz delay and saw all the discussion about the 50Hz api bottle neck. 

However, as I dug into the source code and did several tests, it seems that the major time delay (at least in my case) was not on the RPC side, but rather due to some synchronization problem between the Unity Wrapper and Unity. 

Here is what I have tested: 

1. I modified the setPose() function in Unity wrapper so that it immediately stores the pose passed from the rpc client.

2. In` Unity/AirLibWrapper/Vehicles/MultirotorPawnSimApi.cpp` I created a new overwriting `GetPose()` function that directly returns the pose stored in step 1, instead of fetching and returning pose from Unity. 

3. What I have observed is that the time delay between Set and Get is less than 1ms, so the rpc itself is not the culprit as I have thought. 
![image](https://user-images.githubusercontent.com/39609430/127713960-ee1bafff-bdf6-4f65-9326-6ad35c41e05b.png)

I am not working on this full-time, so it may take me some time to solve this delay problem, but I will keep you updated. : )

Thanks again for the follow up. And I want to thank you personally for your branch 'external_dynamics_simulator_with_example'. It really adds versatility to the airsim.

**Reply by ClarkZha** [reply]
@jonyMarino,

A quick update here: 

I followed the whole pose update pipeline and drew a simplified version of it in the following graph:
![image](https://user-images.githubusercontent.com/39609430/128086290-a90f8518-7553-44e1-acbb-91f2028e2840.png)

When the `simVehiclePoseUpdate()` is called, it will first update the physics engine. Then the new pose given by the physics engine is set to Unity in a second step. 

So there is a delay of 2~3 Unity's loop-time between when the `simVehiclePoseUpdate()` is called and when the new pose is set to Unity. 

I hacked the code so that I can set `poseFromAirLib` in Unity directly with the API. However, I am wondering if there is any reason behind the original API design: i.e. the choice of setting pose to physics and Unity in two consecutive ticks VS setting them simultaneously in the same tick.

**Reply by jonyMarino** [reply]
Thanks, @ClarkZha, for all your investigations and for thanking me! 
When you tested setPose, I think you were already on the server/simulator side. So what you should do is take the time between the call to simSetVehiclePose in your program and when simSetVehiclePose is triggered on the server [here](https://github.com/microsoft/AirSim/blob/6e2512e25d794c9a17c2f45fa94cb441fbc9262b/AirLib/src/api/RpcLibServerBase.cpp#L182).

Regarding the design, the important thing is the physics to be separate from the rendering, nothing more; you can show me your changes if you want.
